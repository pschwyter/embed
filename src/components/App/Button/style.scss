.ada-chaperone-Button {
  width: 44px;
  height: 44px;
  background-color: #23bbda;
  background: none;
  border-radius: 50%;
  position: fixed;
  bottom: 24px;
  right: 24px;
  z-index: 10000;
  cursor: pointer;
  border: 1px solid rgba(black, 0.05);
  transition: all 120ms ease;
  box-shadow: 0 2px 4px rgba(black, 0.1);
  visibility: visible;
  padding: 0;
  margin: 0;
  min-width: initial;
  max-width: initial;
  min-height: initial;
  max-height: initial;
  overflow: hidden;
  background-image: url("https://d124s1zbdqkqqe.cloudfront.net/Dialogue.svg");
  background-size: 28px;
  background-repeat: no-repeat;
  background-position: center;
}
