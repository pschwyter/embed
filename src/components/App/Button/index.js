import { Component } from "preact";
import "./style.scss";

export default class Button extends Component {
  render(props) {
    const { openChat } = props;

    return (
      <button
        title="Open Support Chat"
        accessKey="1"
        tabIndex="0"
        className="ada-chaperone-button"
        onClick={openChat}
      >
        BUTTON
      </button>
    );
  }
}
