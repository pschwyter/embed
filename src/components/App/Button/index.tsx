import classnames from "classnames";
import Client from "models/Client";
import DialogueSvg from "icons/Dialogue.svg";
import SVG from "react-inlinesvg";
import { Component, h } from "preact";
import "./style.scss";

const CHAT_BUTTON_PADDING_SIZE_RATIO = 6.28;
const NOTIFICATION_BADGE_SIZE_RATIO = 0.386;
const NOTIFICATION_BADGE_BOTTOM_RATIO = 0.705;
const NOTIFICATION_BADGE_RIGHT_RATIO = -0.09;
const NOTIFICATION_BADGE_POSITION_RATIO = 0.193;
const NOTIFICATION_BADGE_RADIUS_RATIO = 0.128;

interface InterfaceButton {
  client: Client,
  showIntroEmoji: boolean,
  showNotification: boolean,
  toggleChat(): void
}

export default class Button extends Component<InterfaceButton> {
  state = {
    customIconHasLoaded: false
  };

  setCustomIconHasLoaded = () => {
    this.setState({ customIconHasLoaded: true });
  }

  render(props: InterfaceButton) {
    const { customIconHasLoaded } = this.state;
    const { toggleChat, client, showIntroEmoji, showNotification } = props;

    const useCustomIcon = client.chatButton.icon_type === "custom" || null;
    const chatButtonURL = useCustomIcon ? client.chatButton.icon_path : "static/icons/Dialogue.svg";
    const buttonSize = client.chatButton.size;

    const adaEmbedButtonContainerClassName = classnames("ada-embed-button-container", {
      "ada-embed-button-container--loading": useCustomIcon && !customIconHasLoaded
    });
    const chatButtonIconClassname = classnames("ada-embed-button__icon", {
      "ada-embed-button__icon--hide": showIntroEmoji
    });

    return (
      <div className={adaEmbedButtonContainerClassName}>
        <button
          title="Open Support Chat"
          accessKey="1"
          className="ada-embed-button"
          onClick={toggleChat}
          style={{
            width: buttonSize,
            height: buttonSize,
            backgroundColor: client.chatButton.background_color
          }}
        >
          { useCustomIcon ?
            <SVG
              src={chatButtonURL}
              className={chatButtonIconClassname}
              onLoad={this.setCustomIconHasLoaded}
              cacheGetRequests={true}
            /> :
            <DialogueSvg
              className={chatButtonIconClassname}
            />
          }
          {showIntroEmoji && (
            <img
              alt=""
              role="presentation"
              src={client.intro.body}
              className={
                classnames("ada-embed-button__emoji", {
                  "ada-embed-button__emoji--show": showIntroEmoji
                })
              }
            />
          )}
        </button>
        {showNotification && (
          <svg
            className="ada-embed-notification"
            alt="New Message"
            role="alert"
            style={{
              width: Math.ceil(buttonSize * NOTIFICATION_BADGE_SIZE_RATIO),
              height: Math.ceil(buttonSize * NOTIFICATION_BADGE_SIZE_RATIO),
              bottom: buttonSize * NOTIFICATION_BADGE_BOTTOM_RATIO,
              right: buttonSize * NOTIFICATION_BADGE_RIGHT_RATIO
            }}
          >
              <circle
                className="ada-embed-notification__circle"
                style={{
                  cx: buttonSize * NOTIFICATION_BADGE_POSITION_RATIO,
                  cy: buttonSize * NOTIFICATION_BADGE_POSITION_RATIO,
                  r: buttonSize * NOTIFICATION_BADGE_RADIUS_RATIO,
                  strokeWidth: buttonSize * NOTIFICATION_BADGE_RADIUS_RATIO
                }}
              />
          </svg>
        )}
      </div>
    );
  }
}
