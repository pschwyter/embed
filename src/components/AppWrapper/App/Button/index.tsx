import classnames from "classnames";
import Client from "models/Client";
import DialogueSvg from "icons/Dialogue.svg";
import SVG from "react-inlinesvg";
import { Component, h } from "preact";
import "./style.scss";

const NOTIFICATION_BADGE_SIZE_RATIO = 0.386;
const NOTIFICATION_BADGE_BOTTOM_RATIO = 0.705;
const NOTIFICATION_BADGE_RIGHT_RATIO = -0.09;
const NOTIFICATION_BADGE_BORDER_RATIO = 0.352;

interface InterfaceButton {
  client: Client,
  showIntroEmoji: boolean,
  showNotification: boolean,
  isDraggable: boolean,
  toggleChat(): void
}

export default class Button extends Component<InterfaceButton> {
  state = {
    customIconHasLoaded: false
  };

  setCustomIconHasLoaded = () => {
    this.setState({ customIconHasLoaded: true });
  }

  get notificationStyles() {
    const { client } = this.props;
    const buttonSize = client.chatButton.size;
    const heightAndWidth = Math.ceil(buttonSize * NOTIFICATION_BADGE_SIZE_RATIO);

    return {
      height: heightAndWidth,
      width: heightAndWidth,
      bottom: buttonSize * NOTIFICATION_BADGE_BOTTOM_RATIO,
      right: buttonSize * NOTIFICATION_BADGE_RIGHT_RATIO,
      borderWidth: heightAndWidth * NOTIFICATION_BADGE_BORDER_RATIO
    };
  }

  render(props: InterfaceButton) {
    const { customIconHasLoaded } = this.state;
    const {
      toggleChat, client, showIntroEmoji,
      showNotification, isDraggable
    } = props;

    const useCustomIcon = client.chatButton.icon_type === "custom" || null;
    const chatButtonURL = useCustomIcon ? client.chatButton.icon_path : "static/icons/Dialogue.svg";
    const buttonSize = client.chatButton.size;

    const adaEmbedButtonContainerClassName = classnames("ada-embed-button-container", {
      "ada-embed-button-container--loading": useCustomIcon && !customIconHasLoaded,
      "ada-embed-button-container--not-draggable": !isDraggable
    });
    const chatButtonIconClassname = classnames("ada-embed-button__icon", {
      "ada-embed-button__icon--hide": showIntroEmoji
    });

    return (
      <div className={adaEmbedButtonContainerClassName}>
        <button
          title="Open Support Chat"
          accessKey="1"
          className="ada-embed-button"
          onClick={toggleChat}
          style={{
            width: buttonSize,
            height: buttonSize,
            backgroundColor: client.chatButton.background_color
          }}
        >
          { useCustomIcon ?
            <SVG
              src={chatButtonURL}
              className={chatButtonIconClassname}
              onLoad={this.setCustomIconHasLoaded}
              cacheGetRequests={true}
            /> :
            <DialogueSvg
              className={chatButtonIconClassname}
            />
          }
          {showIntroEmoji && (
            <img
              alt=""
              role="presentation"
              src={client.intro.body}
              className={
                classnames("ada-embed-button__emoji", {
                  "ada-embed-button__emoji--show": showIntroEmoji
                })
              }
            />
          )}
        </button>
        {showNotification && (
          <div
            className="ada-embed-notification"
            alt="New Message"
            role="alert"
            style={this.notificationStyles}
          />
        )}
      </div>
    );
  }
}
