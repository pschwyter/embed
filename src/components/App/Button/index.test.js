/* eslint-env node, jest */

import "jsdom-global/register";
import { shallow } from "preact-render-spy";
import Button from "./index";

function setup(propsOverride = {}) {
  const defaultsProps = {
    client: {
      tint: "#000",
      intro: {
        body: "http://test.com/yolo.svg"
      },
      chatButton: {
        size: 72,
        icon_path: "https://static.ada.support/images/286ca5d2-5311-467a-a5f5-5051ad710db4.svg",
        icon_type: "default",
        background_color: "#3ED1FF"
      }
    }
  };

  const props = Object.assign(defaultsProps, propsOverride);
  const PRSWrapper = shallow(<Button {...props} />);

  return {
    props,
    PRSWrapper
  };
}

describe("<Button />", () => {
  it("should match snapshot", () => {
    const { PRSWrapper } = setup();
    expect(PRSWrapper).toMatchSnapshot();
  });

  it("should show the intro emoji, as well as hide the default icon, when showIntroEmoji is set to true", () => {
    const { PRSWrapper } = setup({
      showIntroEmoji: true
    });

    expect(PRSWrapper.find(".ada-embed-button__emoji").length).toBe(1);
    expect(PRSWrapper.find(".ada-embed-button__icon--hide").length).toBe(1);
  });

  it("it should have the background colour set to the chatButton backgroundColor", () => {
    const { PRSWrapper } = setup();

    expect(PRSWrapper.find(".ada-embed-button").attr("style")["backgroundColor"]).toEqual("#3ED1FF");
  });

  it("should have an accessKey of 1", () => {
    const { PRSWrapper } = setup();

    expect(PRSWrapper.find(".ada-embed-button").attr("accessKey")).toEqual("1");
  });

  it("should show the notification icon when showNotification is set to true", () => {
    const { PRSWrapper } = setup({
      showNotification: true
    });

    expect(PRSWrapper.find(".ada-embed-notification").length).toBe(1);
  });

  it("should not have class properties if isDraggable is true", () => {
    const { PRSWrapper } = setup({isDraggable: true});

    expect(PRSWrapper.find(".ada-embed-button-container--not-draggable").length).toBe(0);
  });
});
