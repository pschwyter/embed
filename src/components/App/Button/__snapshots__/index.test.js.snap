// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should match snapshot 1`] = `
preact-render-spy (1 nodes)
-------
<div class="ada-embed-button-container ada-embed-button-container--not-draggable">
  <button
    title="Open Support Chat"
    accessKey="1"
    class="ada-embed-button"
    style="width: 72px; height: 72px; background-color: #3ED1FF;"
  >
    <undefined class="ada-embed-button__icon"></undefined>
  </button>
</div>

`;
